gen/src/main.rs

```diff
@@ -0,0 +1,130 @@
+extern crate getopts;
+
+extern crate futures;
+extern crate hyper;
+extern crate tokio_core;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate serde_yaml;
+
+#[macro_use]
+extern crate lazy_static;
+
+#[macro_use]
+extern crate matches;
+extern crate regex;
+
+mod download;
+mod generate;
+
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+use download::download;
+use generate::generate_all;
+
+use getopts::Options;
+
+fn main() {
+    let mut opts = Options::new();
+
+    opts.optflag("h", "help", "Print this help menu.");
+    opts.optopt(
+        "u",
+        "unicode-version",
+        "The version of the unicode standard to download.  \
+         If omitted, the local cache will be used.",
+        "VERSION",
+    );
+    opts.optmulti(
+        "",
+        "crate",
+        "A crate to generate tables for.  \
+         This propogates through the tree: if you specify only `unic`, all crates will be built.",
+        "CRATE",
+    );
+
+    let matches = opts.parse(env::args().skip(1))
+        .unwrap_or_else(|e| panic!(e.to_string()));
+
+    if matches.opt_present("h") {
+        print_usage(&opts);
+        return;
+    }
+
+    matches.opt_str("u").map(|version| {
+        println!("Downloading Unicode resources version {}...", version);
+        download(version.as_str()).expect("Failed to download Unicode resources");
+        println!("Finished.");
+    });
+
+    let crates = expand_sub_crates(matches.opt_strs("crate").into_iter());
+
+    if crates.is_empty() {
+        if !matches.opt_present("u") {
+            print_usage(&opts);
+        }
+        return;
+    }
+
+    println!("Generating sources for {} crates...", crates.len());
+    generate_all(crates).expect("Failed to generate sources");
+    println!("Finished.");
+}
+
+/// Path to crate disassembly mapping (YAML).
+///
+/// Format:
+///
+/// ```yaml
+/// ---
+/// crate:
+///   - "crate-subcrate1"
+///   - "crate-subcrate2"
+/// crate-subcrate1:
+///   - "crate-subcrate1-subcrate"
+/// ```
+const CRATES: &'static str = "gen/config/crates.yaml";
+
+/// Take a crate name in the UNIC hierarchy and expand it into all subcrates.
+///
+/// Reads `config/crates.yaml` for the crate disassembly process.
+///
+/// Technically, a crate only needs to be in that mapping if it needs to have sources built.
+/// For consistency, all crates that need generated sources or not should be passed into the
+/// generation module, which will ignore crates it doesn't need to do anything for.
+fn expand_sub_crates<I>(list: I) -> Vec<String>
```

Comment:
If we want to support specifying individual crate-level downloads and generation, here's my rough draft of a config and resolution path to work property:

```text
{
  ucd: {
    version: "10.0.0"
    resources: [
        { url: "http://www.unicode.org/Public/{version}/ucd/ReadMe.txt"
          dest: "data/ucd/ReadMe.txt" }
        { url: "http://www.unicode.org/Public/{version}/ucd/UnicodeData.txt"
          dest: "data/ucd/UnicodeData.txt" }
    ]
    subcrates: {
      core: []
      age: [ { url: "http://www.unicode.org/Public/{}/ucd/DerivedAge.txt"
               dest: "data/ucd/DerivedAge.txt" } ]
    }
  }
}
```

(and so on)

With this config file layout (in whatever format we go with; currently our target is TOML), the valid arguments for `--target` would be:

- Omitted (over everything)
- `ucd` (over all children of `ucd`)
- `ucd-core` (only `ucd-core` and its parents)
- `ucd-age` (only `ucd-age` and its parents)

Resolution would be the following:

- Split `--target` argument over `-`
- If length one
  - Tell that generate module to generate all subcomponents
- Else
  - Tell the generate module specified by the first part to generate subcomponents matching the latter parts

So, `--target ucd` would call (hypothetical) `generate::generate_all()` and `--target ucd-age --target ucd-core` would call `generate::generate_modules(["age", "core"])`.

Downloads would be similarly chunked based on the configuration file.

Once more, let me state that the solution in the above comment is much simpler and (I think) covers all use cases once the transition to the simplified tables has been done. It also has less moving parts and things that need to be changed when adding table generation for a new component.

Also, it's much simpler to handle the CLI arguments.

I'm perfectly willing to shoulder the transition period and the problem of only updating some tables in the git commit. I've already `git blame`d myself for all of the generated tables during this PR 😜